### PR TITLE
[12.0][MIG] sale_exception

### PR DESCRIPTION
**Pull request by sebalix**
Supersedes #812 as it seems inactive but it is the same PR + minor fixes + the new README structure.

### CODE REVIEW

**Comment by sebalix** [comment]
ping @grindtildeath

File: sale_exception/models/sale.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    @api.onchange('order_line')
+    def onchange_ignore_exception(self):
+        if self.state == 'sale':
+            self.ignore_exception = False
+
+    @api.multi
+    def action_confirm(self):
+        if self.detect_exceptions():
+            return self._popup_exceptions()
+        else:
```
Comment (nikul-serpentcs) [inline sale_exception/models/sale.py]: @rvalyi  no need to `else`
```
if self.detect_exceptions():
    return self._popup_exceptions()
return super(SaleOrder, self).action_confirm()
```

**Review by nikul-serpentcs — CHANGES_REQUESTED** [review]
Improved Code

**Review by grindtildeath — COMMENTED** [review]
LGTM :+1:

File: sale_exception/models/sale.py (hunk 1 above)
Comment (sebalix) [reply sale_exception/models/sale.py]: updated

**Review by grindtildeath — APPROVED** [review]

**Comment by tbaden** [comment]
@nikul-serpentcs changes were made, anything else or good to ?

File: sale_exception/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# Copyright 2011 Akretion, Camptocamp, Sodexis
+# Copyright 2018 Akretion, Camptocamp
```
Comment (Tonow-c2c) [inline sale_exception/__manifest__.py]:
```suggestion
# Copyright 2019 Akretion, Camptocamp
```

File: sale_exception/data/sale_exception_data.xml [hunk 3]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (Tonow-c2c) [inline sale_exception/data/sale_exception_data.xml:1]: Licence ?

File: sale_exception/wizard/sale_exception_confirm_view.xml [hunk 4]
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (Tonow-c2c) [inline sale_exception/wizard/sale_exception_confirm_view.xml:1]:
```suggestion
<?xml version="1.0" ?>
```

File: sale_exception/data/sale_exception_data.xml (hunk 3 above)
Comment (Tonow-c2c) [reply sale_exception/data/sale_exception_data.xml:1]:
```suggestion
<?xml version="1.0" ?>
```

File: sale_exception/tests/test_sale_exception.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+class TestSaleException(TestSaleOrder):
+
+    def test_sale_order_exception(self):
+        self.sale_exception_confirm = self.env['sale.exception.confirm']
+        exception = self.env.ref('sale_exception.excep_no_zip')
+        exception.active = True
+        partner = self.env.ref('base.res_partner_1')
+        partner.zip = False
+        p = self.env.ref('product.product_product_6')
+        so1 = self.env['sale.order'].create({
```
Comment (simahawk) [inline sale_exception/tests/test_sale_exception.py:24]: use tracking_disable to save some computation

**Review by simahawk — APPROVED** [review]
LG, just a small change to improve tests if you have time

File: sale_exception/wizard/sale_exception_confirm_view.xml (hunk 4 above)
Comment (sebalix) [reply sale_exception/wizard/sale_exception_confirm_view.xml:1]: @Tonow-c2c I don't think we should change the encoding here, it's relevant only on Python 3 files, no?

File: sale_exception/tests/test_sale_exception.py (hunk 5 above)
Comment (sebalix) [reply sale_exception/tests/test_sale_exception.py:24]: updated (done in `setUpClass`).

**Review by Tonow-c2c — APPROVED** [review]

File: sale_exception/__init__.py [hunk 6]
```diff
@@ -0,0 +1,2 @@
+from . import models
```
Comment (alexey-pelykh) [inline sale_exception/__init__.py:2]: License needed

File: sale_exception/models/__init__.py [hunk 7]
```diff
@@ -0,0 +1 @@
+from . import sale
```
Comment (alexey-pelykh) [inline sale_exception/models/__init__.py:2]: License needed as per template

File: sale_exception/models/sale.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        })
+        return res
+
+    def _sale_get_lines(self):
+        self.ensure_one()
+        return self.order_line
+
+    @api.model
+    def _get_popup_action(self):
+        action = self.env.ref('sale_exception.action_sale_exception_confirm')
```
Comment (alexey-pelykh) [inline sale_exception/models/sale.py]: Maybe just return w/o temp variable?

File: sale_exception/tests/__init__.py [hunk 9]
```diff
@@ -0,0 +1,4 @@
+# (c) 2015 Oihane Crucelaegui - AvanzOSC
```
Comment (alexey-pelykh) [inline sale_exception/tests/__init__.py]: Copyright not needed here

File: sale_exception/views/sale_view.xml [hunk 10]
```diff
@@ -0,0 +1,81 @@
+<?xml version="1.0" ?>
+<!-- License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). -->
+<odoo>
+        <record id="action_sale_test_tree" model="ir.actions.act_window">
```
Comment (alexey-pelykh) [inline sale_exception/views/sale_view.xml]: It would be great to tidy up formatting

File: sale_exception/wizard/__init__.py [hunk 11]
```diff
@@ -0,0 +1 @@
+from . import sale_exception_confirm
```
Comment (alexey-pelykh) [inline sale_exception/wizard/__init__.py:1]: License as per template

**Review by alexey-pelykh — CHANGES_REQUESTED** [review]
Minor code tweaks needed

File: sale_exception/models/__init__.py (hunk 7 above)
Comment (bealdav) [reply sale_exception/models/__init__.py:2]: IMO license in init file is useless. Yes in template, but that's not a real problem. Just my opinion

**Review by ivantodorovich — APPROVED** [review]
LGTM 👍

**Comment by sebalix** [comment]
@alexey-pelykh updated

File: sale_exception/tests/__init__.py [hunk 12]
```diff
@@ -1,4 +1 @@
-# (c) 2015 Oihane Crucelaegui - AvanzOSC
-# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
```
Comment (alexey-pelykh) [inline sale_exception/tests/__init__.py]: License should be here but copyright shouldn't, see https://github.com/OCA/maintainer-tools/tree/master/template/module/

File: sale_exception/models/__init__.py (hunk 7 above)
Comment (alexey-pelykh) [reply sale_exception/models/__init__.py:2]: We still should follow common practices

File: sale_exception/wizard/__init__.py (hunk 11 above)
Comment (alexey-pelykh) [reply sale_exception/wizard/__init__.py:1]: https://github.com/OCA/maintainer-tools/tree/master/template/module/

File: sale_exception/tests/__init__.py (hunk 12 above)
Comment (sebalix) [reply sale_exception/tests/__init__.py]: Fixed

File: sale_exception/models/__init__.py (hunk 7 above)
Comment (sebalix) [reply sale_exception/models/__init__.py:2]: updated

File: sale_exception/views/sale_view.xml (hunk 10 above)
Comment (sebalix) [reply sale_exception/views/sale_view.xml]: done in a separate commit to backport it easily if needed

**Review by alexey-pelykh — APPROVED** [review]

**Comment by sebalix** [comment]
ping @pedrobaeza are you available to merge this one please?

**Review by guewen — APPROVED** [review]

**Pull request merged**